File: lab1/udprecv.c

//  udprecv.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>

#define BUFLEN 512      //Max length of buffer
#define PORT 8889       //The port on which to listen for incoming data
#define LOCAL_LOOP	"127.0.0.1"

void die(char *s)
{
        perror(s);
        exit(1);
}

int main(void)
{
  struct sockaddr_in si_me, si_other;
  int s, i, slen = sizeof(si_other),rlen;
  char buf[BUFLEN];
  //create a UDP socket
  if ((s=socket(AF_INET,SOCK_DGRAM,IPPROTO_UDP))==-1) {die("socket");}
  // zero out the structure
  memset((char *) &si_me, 0, sizeof(si_me));
  si_me.sin_family = AF_INET;
  si_me.sin_port = htons(PORT);
  //si_me.sin_addr.s_addr = htonl(INADDR_ANY);
  si_me.sin_addr.s_addr = inet_addr(LOCAL_LOOP);
  //bind socket to port
  if( bind(s,(struct sockaddr*)&si_me,sizeof(si_me))==-1) {die("bind");}
  //keep listening for data
  while(1)
    {
    printf("Waiting for data...");
    //try to receive some data, this is a blocking call
    if((rlen=recvfrom(s,buf,BUFLEN,0,(struct sockaddr *)&si_other,&slen))==-1)
      {die("recvfrom()"); }
    //print details of the client/peer and the data received
    printf("IP, port %s:%d\n",inet_ntoa(si_other.sin_addr),
				      ntohs(si_other.sin_port));
    printf("Received data: %s\n", buf);
    }
  close(s);
  return 0;
}

File: lab4/udprecv.py
#!/usr/bin/env python3
# udprecv_exact.py - Closer match to C behavior

import socket
import sys

BUFLEN = 512
PORT = 8889
LOCAL_LOOP = "127.0.0.1"
LOCAL_IP = "192.168.2.2"

def die(message):
    print(f"Error: {message}", file=sys.stderr)
    sys.exit(1)

def main():
    # Create UDP socket
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    except socket.error:
        die("socket")
    
    # Configure socket to use specific local address (not INADDR_ANY)
    local_ip = LOCAL_IP
    local_port = PORT
    
    # Bind socket
    try:
        s.bind((local_ip, local_port))
        print(f"Bound to {local_ip}:{local_port}")
    except socket.error:
        die("bind")
    
    print("Waiting for data...")
    
    # Keep listening for data
    try:
        while True:
            # Receive data (blocking call)
            try:
                data, addr = s.recvfrom(BUFLEN)
            except socket.error:
                die("recvfrom()")
            
            # Print client details
            client_ip, client_port = addr
            print(f"IP, port {client_ip}:{client_port}")
            
            # Process received data
            # Handle null terminator like C strings
            null_pos = data.find(b'\0')
            if null_pos != -1:
                # Extract data up to null terminator
                message_data = data[:null_pos]
            else:
                message_data = data
            
            try:
                message = message_data.decode('utf-8')
                print(f"Received data: {message}")
            except UnicodeDecodeError:
                print(f"Received data (raw): {data.hex()}")
                
    except KeyboardInterrupt:
        print("\nServer shutdown.")
    finally:
        s.close()

if __name__ == "__main__":
    main()

